lib/livebook/file_system.ex

```diff
@@ -0,0 +1,133 @@
+defprotocol Livebook.FileSystem do
+  @moduledoc false
+
+  # This protocol defines an interface for file systems
+  # that can be plugged into Livebook.
+
+  @typedoc """
+  A path uniquely idenfies file in the file system.
+
+  Path has most of the semantics of regular file paths,
+  with the following exceptions:
+
+    * path must be be absolute for consistency
+
+    * directory path must have a trailing slash, whereas
+      regular file path must not have a trailing slash.
+      Rationale: some file systems allow a directory and
+      a file with the same name to co-exist, while path
+      needs to distinguish between them
+  """
+  @type path :: String.t()
+
+  @typedoc """
+  A human-readable error message clarifying the operation
+  failure reason.
+  """
+  @type error :: String.t()
+
+  @type access :: :read | :write | :read_write | :none
+
+  @doc """
+  Returns the default directory path.
+
+  To some extent this is similar to current working directory
+  in a regular file system. For most file systems this
+  will just be the root path.
+  """
+  @spec default_path(t()) :: path()
+  def default_path(file_system)
+
+  @doc """
+  Returns a list of files located in the given directory.
+  """
+  @spec list(t(), path(), boolean()) :: {:ok, list(path())} | {:error, error()}
+  def list(file_system, path, recursive)
```

Comment:
What does the recursive flag do in this case?

Reply:
It traverses the directory recursively and returns all paths. We need this in `FileSystem.File.copy` when copying between different file systems. For the local file system this means multiple `ls` calls, but for S3 it's just a single request either way, that's why I made it a flag.

Reply:
Corresponding tests:

https://github.com/livebook-dev/livebook/blob/5edbc63c040ea1bb7c06a843654df54592c9b8f2/test/livebook/file_system/local_test.exs#L38-L88